tools/get-upload-stats.py

```diff
@@ -0,0 +1,580 @@
+        return filter_commits(tags, from_dt, to_dt)
+
+    def get_assets(self, commit: CommitInfo) -> Iterator[AssetInfo]:
+        repo = AnnexRepo(str(self.path))
+        for p, info in repo.get_content_annexinfo(ref=commit.committish).items():
+            relpath = p.relative_to(self.path)
+            if info.get("type") == "file" and relpath.parts[0] not in IGNORED:
+                path = str(PurePosixPath(relpath))
+                subject: Optional[str]
+                if m := re.fullmatch(r"sub-([^/]+)/[^/]+\.nwb", path):
+                    subject = m.group(1)
+                else:
+                    subject = None
+                if info.get("backend"):
+                    # Annexed
+                    yield AssetInfo(
+                        path=path,
+                        size=info["bytesize"],
+                        key=info["key"],
+                        subject=subject,
+                    )
+                else:
+                    # Not annexed
+                    yield AssetInfo(
+                        path=path,
+                        size=int(self.readgit("cat-file", "-s", info["gitshasum"])),
+                        key=info["gitshasum"],
+                        subject=subject,
+                    )
+
+    def get_dandiset_metadata(self, commit: CommitInfo) -> dict:
+        return cast(
+            dict,
+            YAML(typ="safe").load(
+                self.readgit("show", f"{commit.committish}:dandiset.yaml")
+            ),
+        )
+
+    def get_asset_metadata(self, commit: CommitInfo) -> Optional[List[dict]]:
+        assets = json.loads(
+            self.readgit("show", f"{commit.committish}:.dandi/assets.json")
+        )
+        if assets and not (isinstance(assets[0], dict) and "asset_id" in assets[0]):
+            return None
+        else:
+            return [a["metadata"] for a in assets]
+
+    def cmp_commit_assets(
+        self, commit1: CommitInfo, commit2: CommitInfo
+    ) -> CommitDelta:
+        asset_sizes: Dict[str, int] = {}
+        key_qtys: List[Dict[str, int]] = []
+        subjects_sets: List[Set[str]] = []
+        metadata_summaries: List[MetadataSummary] = []
+        for cmt in [commit1, commit2]:
+            keys: Dict[str, int] = Counter()
+            subjects: Set[str] = set()
+            for asset in self.get_assets(cmt):
+                asset_sizes[asset.key] = asset.size
+                keys[asset.key] += 1
+                if asset.subject is not None:
+                    subjects.add(asset.subject)
+            key_qtys.append(keys)
+            subjects_sets.append(subjects)
+            metadata_summaries.append(
+                MetadataSummary.from_metadata(
+                    self.get_dandiset_metadata(cmt),
+                    self.get_asset_metadata(cmt),
+                )
+            )
+        keys1, keys2 = key_qtys
+        added_keys = keys2.keys() - keys1.keys()
```

Comment:
note to myself: damn -- didn't know `dict.keys()` can do that.

Reply:
@yarikoptic It's because `dict.keys()` returns a [dictionary view](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects), which is set-like, so you can do set operations on it.